---
title: Week 3
author: Judit Kisistók
date: '2018-02-14'
slug: week-3
categories: []
tags: []
---



<div id="a.-table-of-contents" class="section level3">
<h3><a name="toc"></a>A. Table of contents</h3>
<ol style="list-style-type: decimal">
<li><a href="#toc">Table of contents</a></li>
<li><a href="#library">Loading libraries</a></li>
<li><a href="#problem9">Problem 9</a></li>
<li><a href="#exercises">Exercises</a>
<ul>
<li><a href="#exercise1">Reproducing in R the Figure 4.1-1 &amp; important details of histogram drawing in R</a></li>
<li><a href="#exercise2">Generating Figure 4.1-2</a></li>
<li><a href="#exercise3">Reproducing Figures 4.1-3 and 4.1-4</a></li>
<li><a href="#exercise4">Trying and evalutating the +/- 2 SEs rule of thumb</a></li>
</ul></li>
</ol>
</div>
<div id="loading-libraries" class="section level3">
<h3><a name="library"></a>Loading libraries</h3>
<pre class="r"><code>library(abd) #we need to load the ABD dataset to be able to access the YeastGenes data in Exercise 1
library(cowplot) #this is just for plotting but you don&#39;t need to worry about it</code></pre>
</div>
<div id="chapter-4-problem-9" class="section level3">
<h3><a name="problem9"></a>Chapter 4, Problem 9</h3>
<p>One of the great discoveries of biology is that organisms have a class of genes called “regulatory genes”, whose only job is to regulate the activity of other genes. How many genes does the typical regulatory gene regulate? A study of interaction networks in yeast (<em>S. cerevisiae</em>) came up with the following data for 109 regulatory genes (Guelzim et al. 2002).</p>
<p>Loading the data and taking a peek at the summary:</p>
<pre class="r"><code>regulatedGenes=read.csv(&quot;chap04q09NumberGenesRegulated.csv&quot;) 
#read.csv() function - use this whenever you have comma-separated data (aka a csv)
summary(regulatedGenes)</code></pre>
<pre><code>##      ngenes      
##  Min.   : 1.000  
##  1st Qu.: 2.000  
##  Median : 6.000  
##  Mean   : 8.312  
##  3rd Qu.:12.000  
##  Max.   :37.000</code></pre>
<div id="what-type-of-graph-should-be-used-to-display-these-data" class="section level4">
<h4>What type of graph should be used to display these data?</h4>
<p>A histogram:</p>
<pre class="r"><code>hist(regulatedGenes$ngenes, breaks=30, col=&quot;cornflowerblue&quot;,
     xlab=&quot;Number of genes regulated&quot;, main=&quot;Regulatory genes in yeast&quot;)</code></pre>
<p><img src="/week3_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
</div>
<div id="what-is-the-estimated-mean-number-of-genes-regulated-by-a-regulatory-gene-in-the-yeast-genome" class="section level4">
<h4>What is the estimated mean number of genes regulated by a regulatory gene in the yeast genome?</h4>
<pre class="r"><code>mean(regulatedGenes$ngenes)</code></pre>
<pre><code>## [1] 8.311927</code></pre>
<pre class="r"><code>median(regulatedGenes$ngenes) #just for reference</code></pre>
<pre><code>## [1] 6</code></pre>
</div>
<div id="what-is-the-standard-error-of-the-mean" class="section level4">
<h4>What is the standard error of the mean?</h4>
<pre class="r"><code>sd(regulatedGenes$ngenes)/sqrt(109)</code></pre>
<pre><code>## [1] 0.7233423</code></pre>
<pre class="r"><code># the standard error is the standard deviation divided by the square root of the sample size and here the sample size is 109</code></pre>
</div>
<div id="explain-what-this-standard-error-measures-and-what-assumptions-are-we-making-when-calculating-it." class="section level4">
<h4>Explain what this standard error measures and what assumptions are we making when calculating it.</h4>
<p>We assume that:</p>
<ul>
<li>the sample is obtained through random sampling<br />
</li>
<li>the statistic is distributed symmetrically</li>
</ul>
<p><em>Take me back to <a href="#toc">the beginning!</a></em></p>
</div>
</div>
<div id="exercises" class="section level3">
<h3><a name="exercises"></a>Exercises</h3>
</div>
<div id="reproducing-the-figure-4.1-1-important-details-of-histogram-drawing-in-r" class="section level3">
<h3><a name="exercise1"></a>Reproducing the Figure 4.1-1 &amp; important details of histogram drawing in R</h3>
<pre class="r"><code>data(HumanGeneLengths) # loading the data of example 4.1 of the book </code></pre>
<div id="take-a-quick-peek-at-the-data-and-try-to-find-the-size-of-the-titin-gene-hint-footnote-on-page-97" class="section level4">
<h4>Take a quick peek at the data and try to find the size of the titin gene (hint: footnote on page 97)</h4>
<pre class="r"><code>summary(HumanGeneLengths)</code></pre>
<pre><code>##   gene.length   
##  Min.   :   60  
##  1st Qu.: 1312  
##  Median : 2226  
##  Mean   : 2622  
##  3rd Qu.: 3444  
##  Max.   :99631</code></pre>
<p>Since we know that the longest human gene is the one encoding titin, we can find its length by looking at the Max value in the summary - 99631 nt.</p>
</div>
<div id="how-many-genes-are-in-the-dataset" class="section level4">
<h4>How many genes are in the dataset?</h4>
<pre class="r"><code>dim(HumanGeneLengths)</code></pre>
<pre><code>## [1] 20290     1</code></pre>
<pre class="r"><code># this indicates the number of rows and the number of columns in the data - hence, the number of genes (20290) and the number of variables (1 - gene length)</code></pre>
</div>
<div id="draw-a-histogram" class="section level4">
<h4>Draw a histogram:</h4>
<pre class="r"><code>hist(HumanGeneLengths$gene.length, breaks=50, col=&quot;deeppink4&quot;, main=&quot;&quot;, xlab=&quot;Gene length (nt)&quot;) </code></pre>
<p><img src="/week3_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
</div>
<div id="identify-what-the-options-col-main-and-xlab-are-allowing-you-to-do." class="section level4">
<h4>Identify what the options col, main and xlab are allowing you to do.</h4>
<p><strong>col:</strong> you can add color to your histogram (<a href="http://www.stat.columbia.edu/~tzheng/files/Rcolor.pdf" target ="_blank">here is a cool color reference guide if you really want to make it fancy</a>)<br />
<strong>main:</strong> adds the title to the histogram<br />
<strong>xlab:</strong> adds the label on the x axis</p>
</div>
<div id="why-is-this-graph-a-bad-display-of-the-data" class="section level4">
<h4>Why is this graph a BAD display of the data?</h4>
<p>Because even though the majority of the data span from 1 to 15000, the histogram above displays the full range up to 10000 - so we get a large empty-looking area. In this case it’s a great idea to say that we want to see a histogram that displays the values between 0 and 15000 - and make a mental note of the “unusual” data point that we disregard in this way.</p>
</div>
<div id="are-there-50-bins-on-the-following-graph-if-not-why-not" class="section level4">
<h4>Are there 50 bins on the following graph? If not, why not?</h4>
<pre class="r"><code># we can save the histogram in a variable so that we can gain some information about it later
histdata &lt;- hist(HumanGeneLengths$gene.length, breaks=50, col=&quot;deeppink4&quot;, main=&quot;&quot;, xlab=&quot;Gene length (nt)&quot;,xlim = c(0,15000)) </code></pre>
<p><img src="/week3_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<pre class="r"><code>histdata</code></pre>
<pre><code>## $breaks
##  [1]      0   2000   4000   6000   8000  10000  12000  14000  16000  18000
## [11]  20000  22000  24000  26000  28000  30000  32000  34000  36000  38000
## [21]  40000  42000  44000  46000  48000  50000  52000  54000  56000  58000
## [31]  60000  62000  64000  66000  68000  70000  72000  74000  76000  78000
## [41]  80000  82000  84000  86000  88000  90000  92000  94000  96000  98000
## [51] 100000
## 
## $counts
##  [1] 8791 7843 2591  738  179   75   30   26    6    5    3    0    0    1
## [15]    0    0    0    0    1    0    0    0    0    0    0    0    0    0
## [29]    0    0    0    0    0    0    0    0    0    0    0    0    0    0
## [43]    0    0    0    0    0    0    0    1
## 
## $density
##  [1] 2.166338e-04 1.932725e-04 6.384919e-05 1.818630e-05 4.411040e-06
##  [6] 1.848201e-06 7.392804e-07 6.407097e-07 1.478561e-07 1.232134e-07
## [11] 7.392804e-08 0.000000e+00 0.000000e+00 2.464268e-08 0.000000e+00
## [16] 0.000000e+00 0.000000e+00 0.000000e+00 2.464268e-08 0.000000e+00
## [21] 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00
## [26] 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00
## [31] 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00
## [36] 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00
## [41] 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00
## [46] 0.000000e+00 0.000000e+00 0.000000e+00 0.000000e+00 2.464268e-08
## 
## $mids
##  [1]  1000  3000  5000  7000  9000 11000 13000 15000 17000 19000 21000
## [12] 23000 25000 27000 29000 31000 33000 35000 37000 39000 41000 43000
## [23] 45000 47000 49000 51000 53000 55000 57000 59000 61000 63000 65000
## [34] 67000 69000 71000 73000 75000 77000 79000 81000 83000 85000 87000
## [45] 89000 91000 93000 95000 97000 99000
## 
## $xname
## [1] &quot;HumanGeneLengths$gene.length&quot;
## 
## $equidist
## [1] TRUE
## 
## attr(,&quot;class&quot;)
## [1] &quot;histogram&quot;</code></pre>
<pre class="r"><code># this way we can see the bins, the counts in each bin, etc.</code></pre>
<p>By examining this, we can see that R indeed gave us 50 bins, but it did so over the span of the entire range - we just can’t see most of them because we chose to display our data up to 15000.</p>
<p>Let’s make a vector of breakpoints for the histogram to enforce the same bins as in the book (500 nt). How many breaks are there?</p>
<pre class="r"><code>myBreaks=seq(from=0, to=15000,  by=500)
length(myBreaks) #number of breaks</code></pre>
<pre><code>## [1] 31</code></pre>
<pre class="r"><code>myBreaks</code></pre>
<pre><code>##  [1]     0   500  1000  1500  2000  2500  3000  3500  4000  4500  5000
## [12]  5500  6000  6500  7000  7500  8000  8500  9000  9500 10000 10500
## [23] 11000 11500 12000 12500 13000 13500 14000 14500 15000</code></pre>
</div>
<div id="why-does-r-choke-here" class="section level4">
<h4>Why does R choke here?</h4>
<pre class="r"><code>#hist(HumanGeneLengths$gene.length, breaks=myBreaks, xlim = c(0,15000))</code></pre>
<p>It says “some ‘x’ not counted; maybe ‘breaks’ do not span range of ‘x’”.</p>
<p>It’s because we’re essentially creating the breaks up to 15000, however, the biggest data point (the gene encoding titin) has the value 99700.</p>
</div>
<div id="what-are-we-doing-here" class="section level4">
<h4>What are we doing here?</h4>
<pre class="r"><code>myBreaks=c(seq(from=0, to=15000, by=500),99700) 
myBreaks</code></pre>
<pre><code>##  [1]     0   500  1000  1500  2000  2500  3000  3500  4000  4500  5000
## [12]  5500  6000  6500  7000  7500  8000  8500  9000  9500 10000 10500
## [23] 11000 11500 12000 12500 13000 13500 14000 14500 15000 99700</code></pre>
<p>We are correcting the previous mistake by adding another break between 15000 and 97500, so that the breaks span the entire range of x.</p>
<p>This one is still bad (but at least more fine-grained):</p>
<pre class="r"><code>hist(HumanGeneLengths$gene.length, breaks=myBreaks, col = &quot;deeppink4&quot;, main = &quot;&quot;, xlab=&quot;Gene length (nt)&quot;)</code></pre>
<p><img src="/week3_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>
</div>
<div id="look-at-the-y-axis-is-this-the-same-y-axis-as-the-book" class="section level4">
<h4>Look at the Y axis: is this the same Y axis as the book?</h4>
<pre class="r"><code>hist(HumanGeneLengths$gene.length, breaks=myBreaks, xlim = c(0,15000), col=&quot;deeppink4&quot;, main=&quot;&quot;, xlab=&quot;Gene length (nt)&quot;, freq=T) </code></pre>
<p><img src="/week3_files/figure-html/unnamed-chunk-15-1.png" width="672" /></p>
<p>It’s not the same Y axis.</p>
</div>
<div id="what-is-the-difference-between-frequency-of-a-class-and-counts-of-observation-per-class" class="section level4">
<h4>What is the difference between frequency of a class and counts of observation per class?</h4>
<pre class="r"><code>hist(HumanGeneLengths$gene.length, breaks=myBreaks, xlim = c(0,15000), col=&quot;deeppink4&quot;, main=&quot;&quot;, xlab=&quot;Gene length (nt)&quot;, freq=F)</code></pre>
<p><img src="/week3_files/figure-html/unnamed-chunk-16-1.png" width="672" /></p>
<p>The first histogram displays counts in the bins and adding them together should sum to the number of observations in total. It shows the <em>absolute frequencies</em>. The second histogram displays the <em>relative frequencies</em> (absolute frequency / total number of observations) and should sum up to 1 - this is still not the exact same scale as we see in the book, however, the goal is not to make an exact carbon copy. This is the best we can do here and the point is really just to understand the difference between absolute and relative frequencies.</p>
<p><em>Take me back to <a href="#toc">the beginning!</a></em></p>
</div>
</div>
<div id="generating-figure-4.1-2" class="section level3">
<h3><a name="exercise2"></a>Generating Figure 4.1-2</h3>
<div id="use-the-r-function-sample-to-get-a-vector-of-100-observations-drawn-from-humangenelengths-and-reproduce-fig-4.1-2" class="section level4">
<h4>Use the R function sample() to get a vector of 100 observations drawn from HumanGeneLengths and reproduce fig 4.1-2</h4>
<p>Hint: Here is a simple example to draw 5 numbers with replacement from a “made-up” empirical distribution - adapt that for your purposes.</p>
<pre class="r"><code>madeup=c(3,3,4,5,6,7,8,2,1,3,5,6,12,435,23)
sample(x = madeup, size = 5,replace = T) # Q:what are the options x=, size= and replace=  specifying here (check the help)?</code></pre>
<pre><code>## [1] 12 23  4  4  7</code></pre>
<p><strong>x:</strong> the data that we want to draw samples from<br />
<strong>size:</strong> the size of the sample that we want to take<br />
<strong>replace:</strong> whether or not we want to sample with replacement</p>
</div>
<div id="discuss-what-is-the-conceptual-difference-between-replace-t-and-replace-f-what-should-we-use-here" class="section level4">
<h4>Discuss what is the conceptual difference between replace = T and replace = F (what should we use here?)</h4>
<p><strong>replace = T</strong>: sampling WITH replacement - when we sample with replacement, the sample values (sampled units) are independent<br />
<strong>replace = F</strong>: sampling WITHOUT replacement - in this case the sample values are not independent, because what we pick as a first value will affect what we can pick as a second value - this complicates computations</p>
<p>As we want a random sample (where each unit has equal chance of being picked and the selection of units is independent), we should sample with replacement.</p>
<pre class="r"><code>hist(sample(x = HumanGeneLengths$gene.length, size = 100,replace = T), 
     breaks=myBreaks, 
     xlim = c(0,15000), 
     col=&quot;deeppink4&quot;, 
     main=&quot;&quot;, 
     xlab=&quot;Gene length (number of nucleotides)&quot;, 
     freq=T) </code></pre>
<p><img src="/week3_files/figure-html/unnamed-chunk-18-1.png" width="672" /></p>
<p>Notice how this histogram is different every single time we run the code - it is due to the randomness in sampling.</p>
<p><em>Take me back to <a href="#toc">the beginning!</a></em></p>
</div>
</div>
<div id="reproducing-figures-4.1-3-and-4.1-4" class="section level3">
<h3><a name="exercise3"></a>Reproducing Figures 4.1-3 and 4.1-4</h3>
<p>Conceptually, to generate by simulation the sampling distribution of mean gene length, we resample 100 datapoints, we calculate the mean in the sample and “write it down” and repeat that operation many times (say 10^4 times).</p>
<p>In most programming languages there is a way to “repeat” a sequence of things x times —&gt; the for loop. Here is one basic example:</p>
<pre class="r"><code>for (year in c(2010,2011,2012,2013,2014,2015)){
  print(paste(&quot;The year is&quot;, year))
}</code></pre>
<pre><code>## [1] &quot;The year is 2010&quot;
## [1] &quot;The year is 2011&quot;
## [1] &quot;The year is 2012&quot;
## [1] &quot;The year is 2013&quot;
## [1] &quot;The year is 2014&quot;
## [1] &quot;The year is 2015&quot;</code></pre>
<p>Here the print statement is executed with various values of the year, but you can also use that to do something just 10 times.</p>
<pre class="r"><code>for (i in 1:10) {
  tryit=sample(x = madeup, size = 5,replace = T) #we take the sample
  mymedian=median(tryit) #we calculate the median of the sample
  cat(&quot;And the median is: &quot;, mymedian, &quot;\n&quot;) #we concatenate and print it out
}</code></pre>
<pre><code>## And the median is:  3 
## And the median is:  3 
## And the median is:  5 
## And the median is:  8 
## And the median is:  6 
## And the median is:  8 
## And the median is:  3 
## And the median is:  5 
## And the median is:  3 
## And the median is:  6</code></pre>
<p>Or we can be smart and instead of “printing it” we “write it down” i.e. store that in a vector:</p>
<pre class="r"><code>myMedians = NULL # this initializes a &quot;nullvector&quot;
for (i in 1:10) {
  tryit=sample(x = madeup, size = 5,replace = T)
  myMedians[i]=median(tryit) #we add the calculated medians to the nullvector
}
myMedians </code></pre>
<pre><code>##  [1] 7 6 3 3 4 5 4 5 1 3</code></pre>
<div id="adapt-this-code-to-generate-fig-4.1-3-and-4.1-4-use-104-samples-to-get-the-sampling-distributions." class="section level4">
<h4>Adapt this code to generate fig 4.1-3 and 4.1-4 (use 10^4 samples to get the sampling distributions).</h4>
<pre class="r"><code>#nullvectors for the means (different sample sizes)
meanLength_100 = NULL
meanLength_20 = NULL
meanLength_500 = NULL

for (i in 1:10000) {
  # taking the samples with different sample sizes
  sample_100 = sample(x = HumanGeneLengths$gene.length, size = 100, replace = T) 
  sample_20 = sample(x = HumanGeneLengths$gene.length, size = 20, replace = T)
  sample_500 = sample(x = HumanGeneLengths$gene.length, size = 500, replace = T)
  # calculating the means
  meanLength_100[i] = mean(sample_100)
  meanLength_20[i] = mean(sample_20)
  meanLength_500[i] = mean(sample_500)
}

par(mfrow=c(3,1)) # creating the 3*1 figure layout

hist(meanLength_20,
     breaks = 80,
     col = &quot;deeppink4&quot;,
     main = &quot;Sample size = 20&quot;,
     xlab = &quot;&quot;, 
     freq = F, 
     xlim = c(1000,5200))

hist(meanLength_100,
     breaks = 80,
     col = &quot;deeppink4&quot;,
     main = &quot;Sample size = 100&quot;,
     xlab = &quot;&quot;, 
     freq = F, 
     xlim = c(1000,5200))

hist(meanLength_500,
     breaks = 80,
     col = &quot;deeppink4&quot;,
     main = &quot;Sample size = 500&quot;,
     xlab = &quot;Sample mean length&quot;, 
     freq = F, 
     xlim = c(1000,5200))</code></pre>
<p><img src="/week3_files/figure-html/unnamed-chunk-22-1.png" width="672" /></p>
<pre class="r"><code># putting it all together on a &quot;fancier&quot; plot
ggplot() + #creates the canvas 
  geom_histogram(mapping = aes(x = meanLength_20, y=..density..), fill = &quot;red&quot;, alpha = 0.6, binwidth = 50, col = &quot;black&quot;) + #adds the first histogram 
  geom_histogram(mapping = aes(meanLength_100, y=..density..), fill = &quot;green&quot;, alpha = 0.6, binwidth = 50, col = &quot;black&quot;) + # adds the second histogram on top of the first
  geom_histogram(mapping = aes(meanLength_500, y=..density..), fill = &quot;deeppink4&quot;, alpha = 0.6, binwidth = 50, col = &quot;black&quot;) + # adds the third histogram on top
  xlab(&quot;Sample mean length&quot;)</code></pre>
<p><img src="/week3_files/figure-html/unnamed-chunk-23-1.png" width="672" /></p>
</div>
<div id="derive-the-sampling-distribution-for-the-sd-in-the-sample-and-the-median." class="section level4">
<h4>Derive the sampling distribution for the SD in the sample and the median.</h4>
<p>Standard deviation:</p>
<pre class="r"><code># this is the exact same code, I&#39;m just calculating the standard deviation instead of the mean

sdLength_100 = NULL
sdLength_20 = NULL
sdLength_500 = NULL

for (i in 1:10000) {
  sample_100_sd = sample(x = HumanGeneLengths$gene.length, size = 100, replace = T)
  sample_20_sd = sample(x = HumanGeneLengths$gene.length, size = 20, replace = T)
  sample_500_sd = sample(x = HumanGeneLengths$gene.length, size = 500, replace = T)
  sdLength_100[i] = sd(sample_100_sd)
  sdLength_20[i] = sd(sample_20_sd)
  sdLength_500[i] = sd(sample_500_sd)
}

par(mfrow=c(3,1))

hist(sdLength_20,
     breaks = 180,
     col = &quot;deeppink4&quot;,
     main = &quot;Sample size = 20&quot;,
     xlab = &quot;&quot;, 
     freq = F)

hist(sdLength_100,
     breaks = 180,
     col = &quot;deeppink4&quot;,
     main = &quot;Sample size = 100&quot;,
     xlab = &quot;&quot;, 
     freq = F)

hist(sdLength_500,
     breaks = 180,
     col = &quot;deeppink4&quot;,
     main = &quot;Sample size = 500&quot;,
     xlab = &quot;Sample standard deviation&quot;, 
     freq = F)</code></pre>
<p><img src="/week3_files/figure-html/unnamed-chunk-24-1.png" width="672" /></p>
<pre class="r"><code>ggplot() +
  geom_histogram(mapping = aes(x = sdLength_20, y=..density..), fill = &quot;red&quot;, alpha = 0.6, binwidth = 30) +
  geom_histogram(mapping = aes(sdLength_100, y=..density..), fill = &quot;green&quot;, alpha = 0.6, binwidth = 30) +
  geom_histogram(mapping = aes(sdLength_500, y=..density..), fill = &quot;deeppink4&quot;, alpha = 0.6, binwidth = 30) +
  xlab(&quot;Sample standard deviation&quot;)</code></pre>
<p><img src="/week3_files/figure-html/unnamed-chunk-25-1.png" width="672" /></p>
<p>Median:</p>
<pre class="r"><code># also the exact same code, just calculating the median

medianLength_100 = NULL
medianLength_20 = NULL
medianLength_500 = NULL

for (i in 1:10000) {
  sample_100_median = sample(x = HumanGeneLengths$gene.length, size = 100, replace = T)
  sample_20_median = sample(x = HumanGeneLengths$gene.length, size = 20, replace = T)
  sample_500_median = sample(x = HumanGeneLengths$gene.length, size = 500, replace = T)
  medianLength_100[i] = median(sample_100_median)
  medianLength_20[i] = median(sample_20_median)
  medianLength_500[i] = median(sample_500_median)
}

par(mfrow=c(3,1))

hist(medianLength_20,
     breaks = 180,
     col = &quot;deeppink4&quot;,
     main = &quot;Sample size = 20&quot;,
     xlab = &quot;&quot;, 
     freq = F)

hist(medianLength_100,
     breaks = 180,
     col = &quot;deeppink4&quot;,
     main = &quot;Sample size = 100&quot;,
     xlab = &quot;&quot;, 
     freq = F)

hist(medianLength_500,
     breaks = 180,
     col = &quot;deeppink4&quot;,
     main = &quot;Sample size = 500&quot;,
     xlab = &quot;Sample median&quot;, 
     freq = F)</code></pre>
<p><img src="/week3_files/figure-html/unnamed-chunk-26-1.png" width="672" /></p>
<pre class="r"><code>ggplot() +
  geom_histogram(mapping = aes(x = medianLength_20, y=..density..), fill = &quot;red&quot;, alpha = 0.6, binwidth = 20, col = &quot;black&quot;) +
  geom_histogram(mapping = aes(medianLength_100, y=..density..), fill = &quot;green&quot;, alpha = 0.6, binwidth = 20, col = &quot;black&quot;) +
  geom_histogram(mapping = aes(medianLength_500, y=..density..), fill = &quot;deeppink4&quot;, alpha = 0.6, binwidth = 20, col = &quot;black&quot;) +
  xlab(&quot;Sample median&quot;)</code></pre>
<p><img src="/week3_files/figure-html/unnamed-chunk-27-1.png" width="672" /></p>
</div>
<div id="compare-the-statistical-precision-we-have-to-estimate-mean-sd-and-median-is-the-summary-statistic-that-entails-most-uncertainty" class="section level4">
<h4>Compare the statistical precision we have to estimate mean SD and median: is the summary statistic that entails most uncertainty?</h4>
<p>So we created all these nice sampling distributions and we can calculate the standard error to get an idea about their uncertainty:</p>
<pre class="r"><code>sd(meanLength_20)/sqrt(20)</code></pre>
<pre><code>## [1] 101.4109</code></pre>
<pre class="r"><code>sd(meanLength_100)/sqrt(100)</code></pre>
<pre><code>## [1] 20.55309</code></pre>
<pre class="r"><code>sd(meanLength_500)/sqrt(500)</code></pre>
<pre><code>## [1] 4.071717</code></pre>
<pre class="r"><code>sd(sdLength_20)/sqrt(20)</code></pre>
<pre><code>## [1] 179.7794</code></pre>
<pre class="r"><code>sd(sdLength_100)/sqrt(100)</code></pre>
<pre><code>## [1] 68.11289</code></pre>
<pre class="r"><code>sd(sdLength_500)/sqrt(500)</code></pre>
<pre><code>## [1] 20.98974</code></pre>
<pre class="r"><code>sd(medianLength_20)/sqrt(20)</code></pre>
<pre><code>## [1] 93.31577</code></pre>
<pre class="r"><code>sd(medianLength_100)/sqrt(100)</code></pre>
<pre><code>## [1] 17.73173</code></pre>
<pre class="r"><code>sd(medianLength_500)/sqrt(500)</code></pre>
<pre><code>## [1] 3.576505</code></pre>
<p>We can see that if we compare the precision of our estimates given the 3 sample sizes, the n = 500 estimate is the most precise which is expected. If we take a look at the different summary statistics at the same sample sizes, we find that the median estimate is the least uncertain - this is also expected given that this statistic is the most resistant to the influence of the outliers while the mean and standard deviation are more sensitive to them (remember, the standard deviation is the most uncertain - it is calculated using the mean, which was already pretty sensitive to extreme observations!).</p>
<p><em>Take me back to <a href="#toc">the beginning!</a></em></p>
</div>
</div>
<div id="trying-and-evaluating-the---2-ses-rule-of-thumb" class="section level3">
<h3><a name="exercise4"></a>Trying and evaluating the +/- 2 SEs rule of thumb</h3>
<div id="generate-20-samples-of-n100-genes-from-the-human-gene-length-distribution-for-each-sample-calculate-the-mean-and-ci-around-the-mean-using-2ses" class="section level4">
<h4>Generate 20 samples of n=100 genes from the human gene length distribution, for each sample calculate the mean and CI around the mean using 2SEs</h4>
<pre class="r"><code>Mean = NULL
SE = NULL
epoch = NULL # don&#39;t worry about this, I just needed it for plotting purposes
sample_size = 100

for(i in 1:20){
  sample &lt;- sample(x = HumanGeneLengths$gene.length, size = sample_size,replace = T)
  Mean[i] = mean(sample)
  SE[i] = sd(sample)/sqrt(sample_size)
  epoch[i] = i
}
# putting it all together in a nice dataframe
df &lt;- data.frame() #creating an empty dataframe
df &lt;- data.frame(cbind(epoch, Mean, SE, Mean-2*SE, Mean+2*SE, sample_size)) #binding the columns together
colnames(df) &lt;- c(&quot;Sample_nr&quot;, &quot;Mean&quot;, &quot;Standard_error&quot;, &quot;Lower_confidence_limit&quot;, &quot;Upper_confidence_limit&quot;, &quot;Sample_size&quot;) #naming the columns in the dataframe

# a fancy plot to visualize the CIs
ggplot() +
  geom_point(data = df, aes(y = Sample_nr, x = Mean)) +
  geom_segment(data = df, aes(x = Lower_confidence_limit,
                              xend = Upper_confidence_limit, 
                              y = Sample_nr,
                              yend = Sample_nr)) +
  geom_vline(xintercept = mean(HumanGeneLengths$gene.length), col = &quot;deeppink4&quot;)</code></pre>
<p><img src="/week3_files/figure-html/unnamed-chunk-29-1.png" width="672" /></p>
</div>
<div id="how-many-intervals-out-of-20-cover-the-true-mean-is-that-what-we-expect-for-a-95-ci-if-you-repeat-that-several-times-numbers-vary-a-bit" class="section level4">
<h4>How many intervals out of 20 “cover” the true mean? Is that what we expect for a 95% CI? (if you repeat that several times, numbers vary a bit)</h4>
<p>It depends - I ran this code several times and sometimes every interval covered the true mean, sometimes only 17 out of 20, etc. This is not exactly what we would expect from a 95% CI, however, the 2SE rule of thumb is just an approximation and keeping this in mind, it works relatively well.</p>
<p>Just to demonstrate how the CIs change when the sample size gets bigger, let’s increase the sample size…</p>
<pre class="r"><code>Mean_1000 = NULL
SE_1000 = NULL
epoch_1000 = NULL
sample_size = 1000


for(i in 1:20){
  sample_1000 &lt;- sample(x = HumanGeneLengths$gene.length, size = sample_size,replace = T)
  Mean_1000[i] = mean(sample_1000)
  SE_1000[i] = sd(sample_1000)/sqrt(1000)
  epoch_1000[i] = 20+i
}
df_1000 &lt;- data.frame(epoch_1000, Mean_1000, SE_1000, Mean_1000-2*SE_1000, Mean_1000+2*SE_1000, sample_size)
colnames(df_1000) &lt;- c(&quot;Sample_nr&quot;, &quot;Mean&quot;, &quot;Standard_error&quot;, &quot;Lower_confidence_limit&quot;, &quot;Upper_confidence_limit&quot;, &quot;Sample_size&quot;)
df &lt;- data.frame(rbind(df, df_1000))

Mean_100000 = NULL
SE_100000 = NULL
epoch_100000 = NULL
sample_size = 100000

for(i in 1:20){
  sample_100000 &lt;- sample(x = HumanGeneLengths$gene.length, size = sample_size,replace = T)
  Mean_100000[i] = mean(sample_100000)
  SE_100000[i] = sd(sample_100000)/sqrt(100000)
  epoch_100000[i] = 40+i
}
df_100000 &lt;- data.frame(epoch_100000, Mean_100000, SE_100000, Mean_100000-2*SE_100000, Mean_100000+2*SE_100000, sample_size)
colnames(df_100000) &lt;- c(&quot;Sample_nr&quot;, &quot;Mean&quot;, &quot;Standard_error&quot;, &quot;Lower_confidence_limit&quot;, &quot;Upper_confidence_limit&quot;, &quot;Sample_size&quot;)
df &lt;- data.frame(rbind(df, df_100000))</code></pre>
<p>Putting it all together:</p>
<pre class="r"><code>ggplot() +
  geom_point(data = df, aes(y = Sample_nr, x = Mean, color = as.factor(df$Sample_size))) +
  geom_segment(data = df, aes(x = Lower_confidence_limit ,
                              xend = Upper_confidence_limit, 
                              y = Sample_nr,
                              yend = Sample_nr,
                              color = as.factor(df$Sample_size))) +
  geom_vline(xintercept = mean(HumanGeneLengths$gene.length)) + 
  ylab(&quot;&quot;) +
  scale_color_discrete(name = &quot;Sample Size&quot;) +
  theme(
  axis.text.y = element_blank(),
  axis.ticks.y = element_blank(),
  axis.line.y = element_blank()) </code></pre>
<p><img src="/week3_files/figure-html/unnamed-chunk-31-1.png" width="672" /></p>
<p>We can see that the bigger the sample size, the narrower the confidence interval - hence, we have more certainty in our estimate.</p>
<p><em>Take me back to <a href="#toc">the beginning!</a></em></p>
</div>
</div>
